python/src/iceberg/expression/literals.py

```diff
@@ -0,0 +1,601 @@
+
+
+class BelowMin(Literal, Singleton):
+    def value(self):
+        raise RuntimeError("BelowMin has no value")
+
+    def to(self, type_var):
+        raise RuntimeError("Cannot change the type of BelowMin")
+
+
+class BooleanLiteral(ComparableLiteral):
+    def __init__(self, value):
+        super(BooleanLiteral, self).__init__(f"BooleanLiteral({value})", value)
+
+    @singledispatchmethod
+    def to(self, type_var):
+        return None
+
+    @to.register(BooleanType)
+    def _(self, type_var):
+        return self
+
+
+class IntegerLiteral(ComparableLiteral):
+    def __init__(self, value):
+        super(IntegerLiteral, self).__init__(f"IntegerLiteral({value})", value)
+
+    @singledispatchmethod
+    def to(self, type_var):
+        return None
+
+    @to.register(IntegerType)
+    def _(self, type_var):
+        return self
+
+    @to.register(LongType)
+    def _(self, type_var):
+        return LongLiteral(self.value)
+
+    @to.register(FloatType)
+    def _(self, type_var):
+        return FloatLiteral(self.value)
+
+    @to.register(DoubleType)
+    def _(self, type_var):
+        return DoubleLiteral(self.value)
+
+    @to.register(DateType)
+    def _(self, type_var):
+        return DateLiteral(self.value)
+
+    @to.register(DecimalType)
+    def _(self, type_var):
+        if type_var.scale == 0:
+            return DecimalLiteral(Decimal(self.value))
+        else:
+            return DecimalLiteral(
+                Decimal(self.value).quantize(
+                    Decimal("." + "".join(["0" for _ in range(1, type_var.scale)]) + "1"), rounding=ROUND_HALF_UP
+                )
+            )
+
+
+class LongLiteral(ComparableLiteral):
+    def __init__(self, value):
+        super(LongLiteral, self).__init__(f"LongLiteral({value})", value)
+
+    @singledispatchmethod
+    def to(self, type_var):
+        return None
+
+    @to.register(LongType)
+    def _(self, type_var):
+        return self
+
+    @to.register(IntegerType)
+    def _(self, type_var):
+        if IntegerType.max < self.value:
+            return AboveMax()
+        elif IntegerType.min > self.value:
+            return BelowMin()
+        return IntegerLiteral(self.value)
+
+    @to.register(FloatType)
+    def _(self, type_var):
+        return FloatLiteral(self.value)
+
+    @to.register(DoubleType)
+    def _(self, type_var):
+        return DoubleLiteral(self.value)
+
+    @to.register(TimeType)
+    def _(self, type_var):
+        return TimeLiteral(self.value)
+
+    @to.register(TimestampType)
+    def _(self, type_var):
+        return TimestampLiteral(self.value)
+
+    @to.register(DecimalType)
+    def _(self, type_var):
+        if type_var.scale == 0:
+            return DecimalLiteral(Decimal(self.value))
+        else:
+            return DecimalLiteral(Decimal(self.value).quantize(Decimal((0, (1,), -type_var.scale)), rounding=ROUND_HALF_UP))
+
+
+class FloatLiteral(ComparableLiteral):
```

Comment:
I think that floats need special implementations of comparison methods like `__eq__`.

The Python VM uses 64-bit double to represent both floating point types. That means that float literals can be more exact than float values that are coming from data files. For example:
| Cell | Python code | Result | Note |
|:-- |:-- |:-- |:-- |
| 1 | `12.34` | `12.34` | Original value
| 2 | `struct.pack('<f', _1)` | `b'\xa4pEA'` | 4-byte binary |
| 3 | `struct.unpack('<f', _2)[0]` | `12.34000015258789` | Lost precision! |
| 4 | `_1 == _3` | `False` | No longer equal |
| 5 | `struct.pack('<d', _1)` | `b'\xaeG\xe1z\x14\xae(@'` | 8-byte binary |
| 6 | `struct.unpack('<d', _5)[0]` | `12.34` | Trailing 0s omitted by repr |
| 7 | `_1 == _6` | `True` | Equal when comparing extra bits |

We need to watch out for the problem shown in [3] for 32-bit float data coming from data files.

One option to fix this is to round-trip the literal value through `struct.pack` and `struct.unpack`. I think that makes the most sense, but it also raises the question: what should we pass into `DoubleLiteral` when converting? My initial take is to either not allow converting a `FloatLiteral` to `DoubleLiteral` or to pass the original python float (64-bit) that was used to create the `FloatLiteral`.

I propose we add `_value32` to this class and use that in overridden comparison methods. That way comparison acts correctly.

```python
class FloatLiteral(Literal):
    def __init__(self, value):
        super(FloatLiteral, self).__init__(f"FloatLiteral({value})", value)
        this._value32 = struct.unpack('<f', struct.pack('<f', value))[0]

    def __eq__(self, value):
        this._value32 == value

    ...
```